kategory/src/main/kotlin/kategory/typeclasses/Monad.kt

```diff
@@ -72,5 +72,56 @@ fun <F, B> Monad<F>.binding(c: suspend MonadContinuation<F, *>.() -> HK<F, B>):
     return continuation.returnedMonad
 }
 
+@RestrictsSuspension
+open class StackSafeMonadContinuation<F, A>(val M: Monad<F>) : Serializable, Continuation<Free<F, A>> {
+
+    override val context = EmptyCoroutineContext
+
+    override fun resume(value: Free<F, A>) {
+        returnedMonad = value
+    }
+
+    override fun resumeWithException(exception: Throwable) {
+        throw exception
+    }
+
+    internal lateinit var returnedMonad: Free<F, A>
+
+    operator suspend fun <B> HK<F, B>.not(): B = this.bind()
+
+    suspend fun <B> HK<F, B>.bind(): B = bind { Free.liftF(this) }
+
+    suspend fun <B> Free<F, B>.bind(): B = bind { this }
```

Comment:
How does name collision work for an HK instance of Free, rather than the child Free class? I'd consider making the extfun based on the HK instead.